### PR TITLE
Feature/scrum 3692

### PR DESCRIPTION
**Pull request by tlm**

### CODE REVIEW

**Comment by tlmiller** [comment]
Have no option to add Jim as reviewer?

File: src/main/java/net/apnic/rdap/error/controller/ErrorController.java [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+    /**
+     * Internal util function for checking if a throwable was caused by the
+     * provided class.
+     *
+     * Rudimentary method for unraveling the wrapping done by Zuul.
+     *
+     * @param th Throwable to check
+     * @param clazz Class to check is contained within the the throwable.
+     */
+    private boolean causedBy(Throwable th, Class<?> clazz)
```
Comment (jimvella) [inline src/main/java/net/apnic/rdap/error/controller/ErrorController.java]: If you just have the Zuul filter generate the error response directly (via composition would be fine), you can avoid coordinating through the wrapped exception and avoid having to introduce the apache dependency,.

File: src/main/java/net/apnic/rdap/error/controller/ErrorController.java [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+     */
+    @ExceptionHandler(Exception.class)
+    public ResponseEntity<RDAPError> handleException(HttpServletRequest request)
+    {
+        String context = request.getRequestURL().toString();
+        return createErrorResponse(HttpStatus.INTERNAL_SERVER_ERROR, context);
+    }
+
+    /**
+     * Handles all Spring exceptions outside of Zuul where no rest handler can
```
Comment (jimvella) [inline src/main/java/net/apnic/rdap/error/controller/ErrorController.java:98]: (Suggestion) This is fine, though you could just map Zuul to /** and deal with this in the Zuul filter.

File: src/main/java/net/apnic/rdap/filter/filters/RDAPPreFilter.java [hunk 3]
```diff
@@ -40,6 +42,8 @@ public Object run()
         RDAPRequestPath path = RDAPRequestPath.createRequestPath(
                 context.getRequest().getRequestURI());
         context.put(RequestContextKeys.RDAP_REQUEST_PATH, path);
+        context.getRequest().setAttribute(ORIGIN_ATT_KEY,
+            context.getRequest().getRequestURL().toString());
```
Comment (jimvella) [inline src/main/java/net/apnic/rdap/filter/filters/RDAPPreFilter.java]: This looks weird. You're taking the request url from the request object and saving it back to the request object with a different name?

File: src/main/java/net/apnic/rdap/rdap/RDAPError.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+/**
+ * Class represent an error response in the RDAP protocol.
+ *
+ * @see https://tools.ietf.org/html/rfc7483
+ */
+public class RDAPError
+    extends RDAPObject
+{
+    private List<String> description = null;
```
Comment (jimvella) [inline src/main/java/net/apnic/rdap/rdap/RDAPError.java]: (suggestion) Do you ever expect to receive this from the client? If not there is an opportunity for immutability here - make all the fields final and ditch the setters.

File: src/main/java/net/apnic/rdap/rdap/RDAPObjectFactory.java [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+/**
+ * Factory helper class for constructing conformant RDAP objects for client
+ * requests that are contextualised and pre filled with default notices.
+ */
+public class RDAPObjectFactory
+{
+    private static final RDAPConformance DEFAULT_CONFORMANCE =
+        RDAPConformance.LEVEL_0;
+
+    private List<RDAPNotice> defaultNotices = null;
```
Comment (jimvella) [inline src/main/java/net/apnic/rdap/rdap/RDAPObjectFactory.java]: (Suggestion) Can you make this immutable (unmodifiable list / immutable RDAPNotice)? then you won't need to copy it.

File: src/main/java/net/apnic/rdap/rdap/RDAPObjectFactory.java [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+     *
+     * @param rdapObject Child object to fill
+     * @return Filled child object
+     */
+    private <T extends RDAPObject> T fillCommonObject(T rdapObject,
+                                                      String context)
+    {
+        return (T)rdapObject
+            .addConformance(DEFAULT_CONFORMANCE)
+            .setNotices(copyDefaultNoticesWithContext(context));
```
Comment (jimvella) [inline src/main/java/net/apnic/rdap/rdap/RDAPObjectFactory.java]: You've got an unchecked cast here since your fluent interface isn't generic. Id just keep it simple:


        rdapObject.addConformance(DEFAULT_CONFORMANCE);
        rdapObject.setNotices(copyDefaultNoticesWithContext(context));
        return rdapObject;

**Comment by jimvella** [comment]
You need to show that given a bad request, the system produces the expected error response. Imo you should implement these tests (specification by example or property based, either is fine) before you produce the implementation.

File: src/main/java/net/apnic/rdap/rdap/RDAPObjectFactory.java (hunk 6 above)
Comment (tlm) [reply src/main/java/net/apnic/rdap/rdap/RDAPObjectFactory.java]: Thanks forgot about these warnings. Will get them fixed up now

File: src/main/java/net/apnic/rdap/filter/filters/RDAPPreFilter.java (hunk 3 above)
Comment (tlm) [reply src/main/java/net/apnic/rdap/filter/filters/RDAPPreFilter.java]: Yeah working around an issue with the way spring is setting up Zuul. It wasn't my first choice doing it this way.

Basically when an error is thrown it gets redirect to the spring error path and you loose all context of the original request. You can try and piece it back together from the error path but there are no certainties made as to correctness which would be worse.

Comment (tlm) [reply src/main/java/net/apnic/rdap/filter/filters/RDAPPreFilter.java]: Most likely this may disappear with your other comments above. Will advice what happens

Comment (apnicjim) [reply src/main/java/net/apnic/rdap/filter/filters/RDAPPreFilter.java]: No worries. Just documenting in the code that it is a work around and why it is there is enough I think.

Comment (tlm) [reply src/main/java/net/apnic/rdap/filter/filters/RDAPPreFilter.java]: Done & Done

File: src/main/java/net/apnic/rdap/rdap/RDAPObjectFactory.java (hunk 6 above)
Comment (tlm) [reply src/main/java/net/apnic/rdap/rdap/RDAPObjectFactory.java]: Done & Done

File: src/main/java/net/apnic/rdap/rdap/RDAPError.java (hunk 4 above)
Comment (tlm) [reply src/main/java/net/apnic/rdap/rdap/RDAPError.java]: How do you mean receive this from the client. This is sent from the server in error cases.

Need some more clarification on what you mean here.

File: src/main/java/net/apnic/rdap/error/controller/ErrorController.java (hunk 1 above)
Comment (tlm) [reply src/main/java/net/apnic/rdap/error/controller/ErrorController.java]: Very true and I was going to do it this way but the following reasons stopped me.

Spring is already there and setup for returning REST response. Doing this in every place that an error could happen would be cumbersome and require a lot manual lifting and writing code for things Spring is doing out of the box already.

On top of that we don't loose this error code it still needs to be there for any exception that happens within Zuul or a runtime exception that has not been caught by the above suggestion.

With this we just end up with 3 places for error handling logic now instead of a central management point thats easier in my opinion to maintain.

The last thought that stopped me from doing it this way originally is its going against the grain of what Zuul is doing. i.e it just a proxy redirect layer to direct traffic and not one that generates tangible responses.

File: src/main/java/net/apnic/rdap/rdap/RDAPError.java (hunk 4 above)
Comment (apnicjim) [reply src/main/java/net/apnic/rdap/rdap/RDAPError.java]: The way POJO deserialization is typically implemented, it requires a no argument constructor and setters. Serialization on the other hand only needs the getters. So if this message is only ever sent (serialized), you can stick final in front of all your members, initialize all the members in the constructor and ditch the setters to make your object immutable and gain all the nice properties that come with that.

File: src/main/java/net/apnic/rdap/rdap/RDAPObjectFactory.java (hunk 5 above)
Comment (tlm) [reply src/main/java/net/apnic/rdap/rdap/RDAPObjectFactory.java]: I have had a look and it gets to tricky the tree structure of objects.

File: src/main/java/net/apnic/rdap/rdap/RDAPError.java (hunk 4 above)
Comment (tlm) [reply src/main/java/net/apnic/rdap/rdap/RDAPError.java]: Ah ok. Happy to implement that but it completely removes the option for staged building of the object which I care more about at the moment. Forces me to not have an object that can not be passed around and be incrementally built.

File: src/main/java/net/apnic/rdap/error/controller/ErrorController.java (hunk 2 above)
Comment (tlm) [reply src/main/java/net/apnic/rdap/error/controller/ErrorController.java:98]: This would break the error handling that spring relies on as nothing would be able to get down to Spring and the paths that it registers. Also slows the server down as Zuul is generating much more expensive errors for things it does not care about.

File: src/main/java/net/apnic/rdap/rdap/RDAPError.java (hunk 4 above)
Comment (apnicjim) [reply src/main/java/net/apnic/rdap/rdap/RDAPError.java]: I think the need to incrementally build is a result of your design of trying to generalize the factory:

from ErrorController:
```
        RDAPError error =
            rdapObjectFactory.createRDAPObject(RDAPError.class, context)
                .setErrorCode(status.value())
                .setTitle(status.getReasonPhrase());

        if(description != null)
        {
            error.addDescription(description);
        }
```

How about just having a component that supplies the common elements, and then just call the constructor?

```
        RDAPConformance conformance = rdapDefaults.getConformance();
        List<RDAPNotice> notices = rdapDefaults.getNotices(context);

        String errorCode = status.value();
        String title = status.getReasonPhrase();
        List<String> descriptions = Optional.ofNullable(description).map(Collections::singletonList).orElse(Collections.emptyList());
        RDAPError rdapError = new RDAPError(errorCode, title, descriptions, notices, conformance);
```

Comment (tlm) [reply src/main/java/net/apnic/rdap/rdap/RDAPError.java]: Thats fine but now everything that ever wants an RDAPError or subclass of RDAPObject now needs to know how to construct not only the child type but the parent type and load it with defaults.

It just seems to me that doing it this way requires any using code now or to come must have knowledge of all the elements that go into constructing a valid RDAPObject for a response.

You could make an RDAPDefaults object that can be passed into the constructor and passed up to the parent object. This method is ok but its still forcing everything that wants to construct an RDAPObject subclass to know about the existence of defaults to construct a valid response.

In my opinion it seems we would just be trading one thing for another and still not realising a perfect design.

Comment (apnicjim) [reply src/main/java/net/apnic/rdap/rdap/RDAPError.java]: I'd make the trade-off for immutability. (and there is no such thing as a perfect design, design is all about making trade-offs).

You could keep your factory by just asking for all the inputs up front - loosing the generic approach by adding a ```createRdapError(...)``` would be the most straight forward.

But I think explicitly putting RDAPDefaults into the constructor is the most cohesive approach - since it tells the reader that RDAPDefaults are actually needed to create a valid instance.

File: src/main/java/net/apnic/rdap/error/controller/ErrorController.java (hunk 1 above)
Comment (apnicjim) [reply src/main/java/net/apnic/rdap/error/controller/ErrorController.java]: The authentication use case is a similar scenario where if the request is not acceptable, the filter immediately returns a response without proxying the request. I don't see how returning an rdap error response is any different. A stack overflow example: https://stackoverflow.com/questions/31099368/netflixoss-zuul-filter-for-rejecting-requests

3 places could easily be one if you just turn the last 2 into POJOs and invoke them from the first filter.

Spring isn't doing that much, you've essentially got the serialization and response headers to worry about.

File: src/main/java/net/apnic/rdap/rdap/RDAPError.java (hunk 4 above)
Comment (tlm) [reply src/main/java/net/apnic/rdap/rdap/RDAPError.java]: See I still disagree that this is the better way to go. We are trying to force immutability into a problem where it may not always be possible.

As an example if I make these structures immutable it means that there now has to be at least one but possibly more combinations of constructors because most if not all properties on some RDAPObjects are optional. So this means either passing null for the args in the constructor or taking different combinations.

On top of that if later on down the track we want to expand this and add more POJO's to encompass more of the possible objects available the factory would now need to implement another custom method just for that type. Or adding another member to an existing object would either require updating the constructor or adding a new constructor which has a knock on effect for all the using code.

Seems to me that adding immutability is trying to solve a problem that doesn't exist and is going to add more code to sections that just need an object. On top of that it starts to add coupling.

As above yes the use of generics has helped force the design into a gradual build approach of objects but I don't think that was the wrong approach. There are multiple parts that need to work on these objects and now they only have to concern themselves with what they care about. Without gradual setters they must take context from other places.

To summarise the RDAP spec specifically is designed as a tree structure for responses. This structure does not lend well to a build it all at once approach as you are going to always have different parts of the code with different concerns.

File: pom.xml [hunk 7]
```diff
@@ -37,6 +37,13 @@
     </dependencyManagement>
 
     <dependencies>
+		<dependency>
```
Comment (JordanShadow) [inline pom.xml]: (suggestion) Keep the indentation in-line with the rest of the dependencies

File: src/main/java/net/apnic/rdap/error/config/ErrorConfiguration.java [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+public class ErrorConfiguration
+{
+    @Bean
+    @Autowired
+    public ZuulErrorFilter zuulErrorFilter(RDAPObjectFactory factory)
+    {
+        return new ZuulErrorFilter(factory);
```
Comment (JordanShadow) [inline src/main/java/net/apnic/rdap/error/config/ErrorConfiguration.java:17]: Google tells me that "zuul" is a genus of dinosaur. I would love for RDAP to have a dinosaur filter, but I suspect that zuul means something different within RDAP. So, what does it mean?

File: src/main/java/net/apnic/rdap/error/controller/ErrorController.java [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.ControllerAdvice;
+import org.springframework.web.bind.annotation.ExceptionHandler;
+import org.springframework.web.servlet.NoHandlerFoundException;
+
+/**
+ * Error controller for spring errors.
+ *
+ * This is used in conjunction with the error controller. Unfotunetly to the way
```
Comment (JordanShadow) [inline src/main/java/net/apnic/rdap/error/controller/ErrorController.java]: (spelling) Unfortunately*

File: src/main/java/net/apnic/rdap/error/config/ErrorConfiguration.java (hunk 8 above)
Comment (JordanShadow) [reply src/main/java/net/apnic/rdap/error/config/ErrorConfiguration.java:17]: Nevermind, found it. https://github.com/Netflix/zuul

File: src/main/java/net/apnic/rdap/error/filters/ZuulErrorFilter.java [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+        {
+            sendErrorResponse(HttpStatus.INTERNAL_SERVER_ERROR, null, context);
+        }
+
+        context.set(SEND_ERROR_FILTER_RAN, true);
+        return null;
+    }
+
+    /**
+     * Constructs and sends a valid RDAP error reponse for a given http status
```
Comment (JordanShadow) [inline src/main/java/net/apnic/rdap/error/filters/ZuulErrorFilter.java]: response*

File: pom.xml (hunk 7 above)
Comment (tlm) [reply pom.xml]: Think this is just a visual mishap in GitHub. Formatting checks out on computer

File: src/main/java/net/apnic/rdap/error/controller/ErrorController.java (hunk 9 above)
Comment (tlm) [reply src/main/java/net/apnic/rdap/error/controller/ErrorController.java]: Done & Done

File: src/main/java/net/apnic/rdap/error/filters/ZuulErrorFilter.java (hunk 10 above)
Comment (tlm) [reply src/main/java/net/apnic/rdap/error/filters/ZuulErrorFilter.java]: Done & Done

File: pom.xml (hunk 7 above)
Comment (JordanShadow) [reply pom.xml]: Ahh cool, no worries then.

**Review by JordanShadow — APPROVED** [review]

File: pom.xml (hunk 7 above)
Comment (GKTheOne) [reply pom.xml]: Tab instead of spaces?

Comment (tlm) [reply pom.xml]: Woah. You are correct. I did check that yesterday but did not get a visual indication tabs where in the file. All fixed now.

File: src/main/java/net/apnic/rdap/rdap/RDAPObjectFactory.java [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+    public List<RDAPNotice> getDefaultNotices()
+    {
+        return defaultNotices;
+    }
+
+
+    /**
+     * Deep copies the default notices
+     */
+    private List<RDAPNotice> copyDefaultNoticesWithContext(String context)
```
Comment (apnicjim) [inline src/main/java/net/apnic/rdap/rdap/RDAPObjectFactory.java]: (Suggestion) Can this be taken out now that this should be immutable / unmodifiable?

Comment (tlm) [reply src/main/java/net/apnic/rdap/rdap/RDAPObjectFactory.java]: Done & Done
